2 features: dbl spend eth detection + syscoinmint json output in deco…

sidhujag opened the pull request:
…deraw

1) We weren't checking to see if ethereum transactions were actually already used to mint. So we create a db to keep track of eth blocknumber/tx index tuple as key and ensure its not already spent. We also need to manage reorgs on the syscoin side and remove these keys from db on syscoin reorg. We have blocknumber as part of the tx mintsyscoin object and blocknumber retrieves the txroot that is checked from the txrootdb, and the txroot is commited to by the block header which is verified by consensus. The path value is an RLP encoded transaction index so we can decode it and get it as a uint32 (not sure what correct format is for the path but i assume uint32 is safe to convert to, seems to pass the test).
2) syscoinmint tx's were not showing json in the "systx" field in decoderawtransaction, just add to the path inside of the mintassettx and detect the type inside.

sidhujag committed:
2 features: dbl spend eth detection + syscoinmint json output in decoderaw

1) We weren't checking to see if ethereum transactions were actually already used to mint. So we create a db to keep track of eth blocknumber/tx index tuple as key and ensure its not already spent. We also need to manage reorgs on the syscoin side and remove these keys from db on syscoin reorg. We have blocknumber as part of the tx mintsyscoin object and blocknumber retrieves the txroot that is checked from the txrootdb, and the txroot is commited to by the block header which is verified by consensus. The path value is an RLP encoded transaction index so we can decode it and get it as a uint32 (not sure what correct format is for the path but i assume uint32 is safe to convert to, seems to pass the test).
2) syscoinmint tx's were not showing json in the "systx" field in decoderawtransaction, just add to the path inside of the mintassettx and detect the type inside.

sidhujag committed:
typos

willyko commented on src/services/assetallocation.cpp:
```diff
@@ -1229,48 +1238,58 @@ bool AssetMintTxToJson(const CTransaction& tx, UniValue &entry){
+            entry.pushKV("asset_guid", (int)mintsyscoin.assetAllocationTuple.nAsset);
+            entry.pushKV("sender", mintsyscoin.assetAllocationTuple.witnessAddress.ToString());
+        }
+        if(tx.nVersion == SYSCOIN_TX_VERSION_ASSET_ALLOCATION_MINT){
```
can be combined with L1255?

willyko replied:
also L1251

sidhujag committed:
remove Dosy checks around txroot validation

we want to avoid blocks from being flagged as invalid and thus avoid reprocessing them as txroot checks are sometimes intermittent (they are dependent on the relayer). So we flag as error instead of invalid which at the tx level it rejects and doesn't ban and on the block level it rejects but reprocesses again (assuming relayer feeds the correct information at some time, will reprocess the block if the transaction was legit).

We should look at this later and figure out when Dosy checks are needed like thresholds where txroots can fall into after looking at some statistics of borderline rejections. Once we have those stats we can then add Dosy checks around transactions that are obviously wrong versus borderline.

willyko commented on src/services/assetconsensus.cpp:
```diff
@@ -1512,3 +1544,137 @@ bool CheckSyscoinLockedOutpoints(const CTransactionRef &tx, CValidationState& st
+    {
+
+        // cutoff to keep blocks is ~3 week of blocks is about 120k blocks
+        cutoffHeight = fNewGethSyncHeight - (MAX_ETHEREUM_TX_ROOTS*3);
```
prevent unsigned int minus unsigned int

sidhujag committed:
update contract for testnet

willyko commented on src/services/assetconsensus.cpp:
```diff
@@ -1512,3 +1544,137 @@ bool CheckSyscoinLockedOutpoints(const CTransactionRef &tx, CValidationState& st
+    vector<uint32_t> vecHeightKeys;
+    uint32_t nKey = 0;
+    uint32_t nKeyIndex = 0;
+    const uint32_t& nKeyCutoff = fGethCurrentHeight - MAX_ETHEREUM_TX_ROOTS;
```
avoid uint - uint

sidhujag committed:
fix int->uint issues around pruning tx roots

sidhujag committed:
cleanup
